avm/src/daemon/linux.rs

```diff
@@ -39,71 +39,99 @@ struct VMStatsV1 {
   let agzb64 = agz_b64.to_string();
   task::spawn(async move {
     let dns = DNS::new("alandeploy.com");
-    let ip_addrs = dns.get_ip_addrs(&app_id).await;
-    let cluster_size = ip_addrs.len();
-    let lrh = LogRendezvousHash::new(ip_addrs);
+    let vms = dns.get_vms(&cluster_id).await;
+    let cluster_size = vms.len();
+    let ips = vms.iter().map(|vm| vm.private_ip_addr.to_string()).collect();
+    let alan_version = &vms[0].alan_version;
+    let lrh = LogRendezvousHash::new(ips);
     let leader_ip = lrh.get_leader_id();
-    let platform = host::platform().await.expect("Failed to get platform information");
-    let self_ip = platform.hostname().replace("ip-", "");
+    let self_ip = get_private_ip().await;
     if leader_ip == self_ip {
       // TODO better period determination
       let period = Duration::from_secs(30);
       loop {
-        let memory = memory::memory().await.expect("Failed to get system memory information");
-        let swap = memory::swap().await.expect("Failed to get swap information");
-        let stats = VMStatsV1 {
-          cpuSecs: cpu::times().map(|r| {
-            let cpu = r.expect("Failed to get CPU times");
-            CPUSecsV1 {
-              user: cpu.user().get::<second>(),
-              system: cpu.system().get::<second>(),
-              idle: cpu.idle().get::<second>(),
-              interrupt: cpu.irq().get::<second>(),
-              nice: cpu.nice().get::<second>(),
-              iowait: cpu.io_wait().get::<second>(),
-            }
-          }).collect().await,
-          totalMemoryKb: memory.total().get::<kilobyte>(),
-          availableMemoryKb: memory.available().get::<kilobyte>(),
-          freeMemoryKb: memory.free().get::<kilobyte>(),
-          activeMemoryKb: memory.active().get::<kilobyte>(),
-          usedMemoryKb: memory.used().get::<kilobyte>(),
-          totalSwapKb: swap.total().get::<kilobyte>(),
-          usedSwapKb: swap.used().get::<kilobyte>(),
-          freeSwapKb: swap.free().get::<kilobyte>(),
-        };
-        
-        let stats_body = json!({
-          "deployToken": deploy_token,
-          "vmStats": stats,
-          "clusterSize": cluster_size,
-        });
-    
-        let resp = post("stats", stats_body).await;
-        let delta: i32 = resp.parse().expect("Failed to parse cluster delta");
+        let delta = post_v1_stats(&deploy_token, cluster_size).await;
+        println!("VM stats sent for cluster {} of size {}. Cluster delta: {}.", cluster_id, cluster_size, delta);
         if delta != 0 {
-          let scale_body = json!({
-            "appId": app_id,
-            "agzB64": agzb64,
-            "deployToken": deploy_token,
-            "clusterDelta": delta,
-          });
-          post("scale", scale_body).await;
+          post_v1_scale(&cluster_id, &agzb64, &deploy_token, alan_version, delta).await;
         }
+        // TODO maybe ask deploy service to /upgrade if delta = 0 and not on the latest alan version
```

Comment:
I don't think so. There's still the regression potentiality. We should only upgrade the `alan` version on user-triggered upgrades.